### DISCUSSION TITLE
About  Collection loading and RAM requirement

### DISCUSSION THREAD

**Discussion by rationalga** [milvus-io/milvus#30128] [Q&A and General discussion] [answered]
Hi, we are using docker based latest standalone Milvus server and   using DiskANN index on a collection. We chose DiskANN because the index otherwise would have been too large to fit in RAM. I am bit confused about the loading of collection. 

- Before searching we need to load the collection in memory but what if the size of the collection is bigger than memory ? i.e the raw vectors inserted into the collection are bigger than the memory. 
- If that is true then DiskANN is only useful if the collection (raw data) itself can fit in RAM?. 
- Is there a way to create collections that are bigger than RAM (any configurations).

Thanks

**Comment by yhmo** [comment]
Milvus requires loading the collection's index into memory before any search/query. 
The data size of a collection can be estimated by: 
1. vector field data size -------------------------- dimension * row_count * 4 Bytes
2. scalar field data size -------------------------- bytes_of_each_value * row_count

Then you create an index for the vector field, different indexes has different sizes:
- FLAT ---------------------------- the size is equal to the original vector field data size
- IVF_FLAT/HNSW ----------- the size is a bit larger than the original vector field data size
- IVF_PQ/IVF_SQ8 ----------- the size is about 24%~30% of the original vector field data size
- DISKANN ---------------------- this is a special index, it has two parts, in-memory index and on-disk data, the in-memory index size is about 1/6~1/4 of the original vector field data size


So, If your collection original data size is 100GB:
- FLAT will require 100GB RAM
- IVF_FLAT/HNSW will require 100GB+ RAM
- IVF_PQ/IVF_SQ8 will require 25~30GB RAM
- DISKANN will require 17~25GB RAM

**Comment by rationalga** [comment]
Thank you very much for detailed answer. If I understood correctly then a collection of 100GB can be loaded on a machine with 32GB RAM when using DISKANN. right?

**Reply by yhmo** [reply, marked as answer]
Theoretically, yes. To avoid potential OOM risk,  we often use a safety ratio to allocate RAM. Typically the safety ratio is 1.5 ~ 2.0. So, we recommend allocating 50GB RAM totally to the query nodes.

**Reply by rationalga** [reply]
Got it. Thank you.
